### PR TITLE
Fixes #1625 - Close stream handle when the current instance loses

### LINKED ISSUE #1625: SSE Stream handle should close the connection, when the leadership is lost
If marathon is setup in a quorum, only the leader should send events.
If one leader abdicates with currently connected stream handlers, those stream handlers should get notified and the stream should be closed. 
The following connection will be established to the current elected leader.

### PR DESCRIPTION
**Pull request by kolloch**
leadership

Also:

* Refactor Leadership management into EventStream messages
* Report the number of open event streams via metrics

### CODE REVIEW

**Comment by kolloch** [comment]
rebased

File: src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala [hunk 1]
```diff
@@ -291,7 +294,7 @@ class MarathonSchedulerService @Inject() (
     leader.set(true)
     runDriver(abdicateOption)
 
-    schedulerActor ! Start
+    eventStream.publish(LeadershipStateEvent.ElectedAsLeader)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala]: This is cool for logic, that gets executed sometimes in the future after the leader was elected. I could think of other logic (think of migration etc), that has to be executed after election but before the driver starts. I know this is not part of this PR, but would be really useful.

File: src/main/scala/mesosphere/marathon/event/Events.scala [hunk 2]
```diff
@@ -40,6 +40,14 @@ object EventModule {
   final val busName = "events"
 }
 
+/** Local leadership events. They are not delivered via the event endpoints. */
+sealed trait LeadershipStateEvent
+
+object LeadershipStateEvent {
+  case object ElectedAsLeader
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/Events.scala]: I could imagine, that this kind of event is also helpful for external services. 
For that it should have something like the IP Address.
As SSE Listener this is would probably be the last message, that is received.

File: src/main/scala/mesosphere/marathon/event/Events.scala [hunk 3]
```diff
@@ -40,6 +40,14 @@ object EventModule {
...
+  case object Standby
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/Events.scala]: Same comment as for ElectedAsLeader.

File: src/main/scala/mesosphere/marathon/event/Events.scala [hunk 4]
```diff
@@ -40,6 +40,14 @@ object EventModule {
   final val busName = "events"
 }
 
+/** Local leadership events. They are not delivered via the event endpoints. */
+sealed trait LeadershipStateEvent
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/Events.scala]: I wonder, if we should make the concept of "Local Events" more visible (like exception hierarchy for checked, runtime exceptions).
See my comments below, where the question is: do we need "Local Events" for this.

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala [hunk 5]
```diff
@@ -16,49 +22,117 @@ trait HttpEventStreamHandle {
   def close(): Unit
 }
 
+class HttpEventStreamActorMetrics @Inject() (metrics: Metrics) {
+  val numberOfStreams: AtomicIntGauge =
+    metrics.gauge(metrics.name("api", getClass, "number-of-streams"), new AtomicIntGauge)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: I would like to have a more specific method like metrics.name -> metrics.api or something like that. So that we don't deal with strings here.

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala [hunk 6]
```diff
@@ -16,49 +22,117 @@ trait HttpEventStreamHandle {
...
+  private[this] val active: Receive = behaviour(acceptingNewConnections)
+  private[this] val standby: Receive = behaviour(rejectingNewConnections)
+
+  /**
+    * Helper method to create behaviours.
+    * The behaviours only differ in how they deal with new connections.
+    */
+  private[this] def behaviour(newConnectionBehaviour: Receive): Receive = {
+    // Prevent fatal warning about deriving type Any as type parameter
+    def orElse(receives: Receive*): Receive = receives.reduceLeft { _.orElse[Any, Unit](_) }
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: Minor: For me the simpler version would be easier to understand: Seq(handleLeadership, cleanupHandlerActors, newConnectionBehaviour, warnAboutUnknownMessages).reduceLeft { _.orElse[Any, Unit](_) }

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala [hunk 7]
```diff
@@ -16,49 +22,117 @@ trait HttpEventStreamHandle {
...
-      eventStream,
-      maxOutstandingMessages), handle.id)
-    context.watch(actor)
-    clients += handle -> actor
+  // behaviour components
+
+  /** Immediately close new connections. */
+  private[this] def rejectingNewConnections: Receive = {
+    case HttpEventStreamConnectionOpen(handle) =>
+      log.warn("Ignoring open connection request. Closing handle.")
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: The LeaderProxyFilter should prevent that from happening, correct?

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala [hunk 8]
```diff
@@ -16,49 +22,117 @@ trait HttpEventStreamHandle {
...
+        s"Current nr of listeners: ${streamHandleActors.size}")
     }
   }
 
-  def removeActor(actor: ActorRef): Unit = {
-    clients.find(_._2 == actor).foreach {
+  private[this] def removeHandlerActor(actor: ActorRef): Unit = {
+    streamHandleActors.find(_._2 == actor).foreach {
       case (handle, ref) =>
         log.error(s"Actor terminated unexpectedly: $handle")
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: This is no longer unexpected, since it is used when this actor becomes Standby. Either change the logic in the standby case to have the logic here as exception (that was my original intend) or change the log statement.

**Comment by aquamatthias** [comment]
This looks generally very good. 
I have some questions regarding local events and some minor comments.  
Can you also look into the tests? ( I restarted it, but it keeps failing )

File: src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala (hunk 1 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala]: Yes, I agree. (It would be useful and it is not part of this PR).

You can create an issue for that.

File: src/main/scala/mesosphere/marathon/event/Events.scala (hunk 2 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/event/Events.scala]: I thought about that but I think it is something different that might be useful on its own.

These are events from the viewpoint of the Marathon instance processing these. They are not global events. You could turn them into global events but it gets more complicated.

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala (hunk 5 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: What about moving `"service"` and `"api"` as constants to an object `mesosphere.marathon.MetricPrefixes`?

I would not like to introduce an "api" method in Metrics since I would like to keep it general.

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala (hunk 6 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: Ok, I'll change that.

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala (hunk 7 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: There is always a race condition.

"Doppelt hält besser ;)"

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala (hunk 8 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: Good point!

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala (hunk 5 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: I agree.

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala (hunk 7 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: I would not omit the code - like double checks.

**Comment by kolloch** [comment]
changed according to your suggestions or our discussion

File: src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala (hunk 5 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/event/http/HttpEventStreamActor.scala]: You introduced the MetricsPrefixes but did not use it here...

**Comment by aquamatthias** [comment]
The one "api" slipped through.
Also the integration tests fail constantly on my local machine.

**Comment by kolloch** [comment]
I used the supplied constants for the "api" string and I fixed the guice wiring.

In a separate commit, I also made ProcessKeeper.startProcess fail immediately if the started process unexpectedly exits.

**Comment by kolloch** [comment]
Integration tests still fail.

**Comment by kolloch** [comment]
I changed the leadership subscription logic, too.

**Comment by aquamatthias** [comment]
Cool!
